File: app/src/main/java/com/bearcave/passageplanning/tides/web/UrlBuilder.java
package com.bearcave.passageplanning.tides.web;

import com.bearcave.passageplanning.tides.web.configurationitems.FileType;
import com.bearcave.passageplanning.tides.web.configurationitems.Gauge;
import com.bearcave.passageplanning.tides.web.configurationitems.MinuteStep;
import com.bearcave.passageplanning.tides.web.configurationitems.NumberOfDays;

import org.joda.time.DateTime;


public class UrlBuilder {

    private static final String URL = "http://tidepredictions.pla.co.uk/listing_page";
    private static final String separator = "/";

    private String fileType;
    private String gaugeCode;
    private DateTime time;
    private int numberOfDays;
    private int step;

    public UrlBuilder(){
        setFileType(FileType.XML);
        setNumberOfDays(NumberOfDays.ONE);
        setStep(MinuteStep.TEN);
    }

    public String build() {
        return URL
                + separator + this.gaugeCode
                + separator + this.time.getYear()
                + separator + this.time.getMonthOfYear()
                + separator + this.time.getDayOfMonth()
                + separator + this.numberOfDays
                + separator + this.step
                + separator + 0
                + separator + this.fileType
                ;
    }

    public void setFileType(FileType fileType) {
        this.fileType = fileType.getTypeName();
    }

    public void setGaugeCode(Gauge gaugeCode) {
        this.gaugeCode = gaugeCode.getCode();
    }

    public void setTime(DateTime time) {
        this.time = time;
    }

    public void setNumberOfDays(NumberOfDays numberOfDays) {
        this.numberOfDays = numberOfDays.getLength();
    }

    public void setStep(MinuteStep step) {
        this.step = step.getStep();
    }
}

File: app/src/main/java/com/bearcave/passageplanning/tides/web/UrlBuilder.kt
package com.bearcave.passageplanning.tides.web

import com.bearcave.passageplanning.tides.web.configurationitems.FileType
import com.bearcave.passageplanning.tides.web.configurationitems.Gauge
import com.bearcave.passageplanning.tides.web.configurationitems.MinuteStep
import com.bearcave.passageplanning.tides.web.configurationitems.NumberOfDays

import org.joda.time.DateTime


class UrlBuilder(
        val gaugeCode: Gauge,
        val time: DateTime? = null,
        val numberOfDays: NumberOfDays = NumberOfDays.ONE,
        val step: MinuteStep = MinuteStep.TEN,
        val fileType: FileType = FileType.XML
) {



    fun build(): String {
        val builder = StringBuilder()
        builder.append(URL)
        builder.append(separator + gaugeCode.code)
        builder.append(separator + time!!.year)
        builder.append(separator + time.monthOfYear)
        builder.append(separator + time.dayOfMonth)
        builder.append(separator + numberOfDays.length)
        builder.append(separator + step.value)
        builder.append(separator + 0)
        builder.append(separator + fileType.typeName)

        return builder.toString()
    }

    companion object {
        val URL = "http://tidepredictions.pla.co.uk/listing_page"
        val separator = "/"
    }
}






